presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -126,695 +122,988 @@ public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, Sym
+                    mapper.getMapping());
         }
 
         @Override
-        public PlanNode visitSort(SortNode node, RewriteContext<Void> context)
+        public PlanAndMappings visitApply(ApplyNode node, UnaliasContext context)
         {
-            PlanNode source = context.rewrite(node.getSource());
+            // it is assumed that apart from correlation (and possibly outer correlation), symbols are distinct between Input and Subquery
+            // rewrite Input
+            PlanAndMappings rewrittenInput = node.getInput().accept(this, context);
+            SymbolMapper mapper = new SymbolMapper(rewrittenInput.getSymbolMappings());
+
+            // rewrite correlation with mapping from Input
+            List<Symbol> rewrittenCorrelation = mapper.mapAndDistinct(node.getCorrelation());
+
+            // extract new mappings for correlation symbols to apply in Subquery
+            Set<Symbol> correlationSymbols = ImmutableSet.copyOf(node.getCorrelation());
+            Map<Symbol, Symbol> correlationMapping = mapper.getMapping().entrySet().stream()
+                    .filter(mapping -> correlationSymbols.contains(mapping.getKey()))
+                    .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+
+            Map<Symbol, Symbol> mappingForSubquery = new HashMap<>();
+            mappingForSubquery.putAll(context.getCorrelationMapping());
+            mappingForSubquery.putAll(correlationMapping);
+
+            // rewrite Subquery
+            PlanAndMappings rewrittenSubquery = node.getSubquery().accept(this, new UnaliasContext(mappingForSubquery));
+
+            // unify mappings from Input and Subquery to rewrite Subquery assignments
+            Map<Symbol, Symbol> resultMapping = new HashMap<>();
+            resultMapping.putAll(rewrittenInput.getSymbolMappings());
+            resultMapping.putAll(rewrittenSubquery.getSymbolMappings());
+            mapper = new SymbolMapper(resultMapping);
+
+            ImmutableList.Builder<Map.Entry<Symbol, Expression>> builder = ImmutableList.builder();
+            for (Map.Entry<Symbol, Expression> assignment : node.getSubqueryAssignments().entrySet()) {
+                builder.add(new SimpleEntry<>(mapper.map(assignment.getKey()), mapper.map(assignment.getValue())));
+            }
+            List<Map.Entry<Symbol, Expression>> rewrittenAssignments = builder.build();
 
-            return new SortNode(node.getId(), source, canonicalizeAndDistinct(node.getOrderingScheme()), node.isPartial());
+            // deduplicate assignments
+            Map<Symbol, Expression> deduplicateAssignments = new HashMap<>();
+            for (Map.Entry<Symbol, Expression> assignment : rewrittenAssignments) {
+                Expression previous = deduplicateAssignments.put(assignment.getKey(), assignment.getValue());
+                checkState(previous == null || assignment.getValue().equals(previous), "different expressions assigned to the same symbol");
```

Comment:
This could be simplified as I described elsewhere

Reply:
same as in `visitProject()`